src/server/server_family.cc

```diff
@@ -1172,10 +1172,29 @@ string GetPassword() {
   return "";
 }
 
+void ServerFamily::SendInvalidationMessages() const {
+  // send invalidation message (caused by flushdb) to all the clients which
+  // turned on client tracking
+  auto cb = [](unsigned thread_index, util::Connection* conn) {
+    facade::ConnectionContext* fc = static_cast<facade::Connection*>(conn)->cntx();
+    if (fc) {
+      ConnectionContext* cntx = static_cast<ConnectionContext*>(fc);
+      if (cntx->conn()->IsTrackingOn()) {
+        facade::Connection::InvalidationMessage x;
+        x.invalidate_due_to_flush = true;
+        cntx->conn()->SendInvalidationMessageAsync(x);
+      }
+    }
+  };
+  for (auto* listener : listeners_) {
+    listener->TraverseConnections(cb);
+  }
+}
+
 void ServerFamily::FlushDb(CmdArgList args, ConnectionContext* cntx) {
   DCHECK(cntx->transaction);
   Drakarys(cntx->transaction, cntx->transaction->GetDbIndex());
```

Comment:
I think we should clear client_tracking_map_ in flush

Reply:
this is not needed as Drakarys will finally end up calling PerformDeletion for each key, which will get the tracking table removed at the end.

Reply:
FushDb calls PerformDeletion only for external keys (ssd).
If we would have called PerformDeletion for all keys in flushdb the client would get the invalidation message twice, one for all keys and one for each key in the database.
If we will have client tracking with tiering I believe the behaviour of flushall will be wrong. This is why I wrote above to get the invalidation message out of PerformDeletion that should just free the data and update table counters

Reply:
i see what you meant, I just added another commit which clear client_tracking_map in FlushDB of DbSlice:

https://github.com/dragonflydb/dragonfly/pull/2280/commits/287c3f54f552e4a426c23b6c1689901107b8ab6d

This shall fix both problem:
1) FlushDB and FlushAll of ServerFamily will end up clearing the client tracking map
2) For SSD tiering, since the map has been cleared before `PerformDeletion` gets called, notification will not be sent out twice.